source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -73,89 +196,165 @@ void Filter::onReadWorker() {
+    if (nread == -1 && errno == EAGAIN) {
+      return false;
+    } else if (nread != bytes_avail) {
+      throw EnvoyException("failed to read proxy protocol extension");
     }
-    socket.setRemoteAddress(remote_address);
+    proxyProtocolHeader_.value().extensions_length -= nread;
   }
-
-  // Release the file event so that we do not interfere with the connection read events.
-  file_event_.reset();
-  cb_->continueFilterChain(true);
+  return true;
 }
 
-bool Filter::readLine(int fd, std::string& s) {
+bool Filter::readProxyHeader(int fd) {
   while (buf_off_ < MAX_PROXY_PROTO_LEN) {
-    ssize_t nread = recv(fd, buf_ + buf_off_, MAX_PROXY_PROTO_LEN - buf_off_, MSG_PEEK);
+    int bytes_avail;
+    auto& os_syscalls = Api::OsSysCallsSingleton::get();
+
+    if (os_syscalls.ioctl(fd, FIONREAD, &bytes_avail) < 0) {
+      throw EnvoyException("failed to read proxy protocol (no bytes avail)");
+    }
+
+    if (bytes_avail == 0) {
+      return false;
+    }
+
+    bytes_avail = std::min(size_t(bytes_avail), MAX_PROXY_PROTO_LEN - buf_off_);
+
+    ssize_t nread = os_syscalls.recv(fd, buf_ + buf_off_, bytes_avail, MSG_PEEK);
 
     if (nread == -1 && errno == EAGAIN) {
       return false;
     } else if (nread < 1) {
-      throw EnvoyException("failed to read proxy protocol");
+      throw EnvoyException("failed to read proxy protocol (no bytes read)");
     }
 
-    bool found = false;
-    // continue searching buf_ from where we left off
-    for (; search_index_ < buf_off_ + nread; search_index_++) {
-      if (buf_[search_index_] == '\n' && buf_[search_index_ - 1] == '\r') {
-        search_index_++;
-        found = true;
-        break;
+    if (buf_off_ + nread >= PROXY_PROTO_V2_HEADER_LEN) {
+      const char* sig = PROXY_PROTO_V2_SIGNATURE;
+      if (!memcmp(buf_, sig, PROXY_PROTO_V2_SIGNATURE_LEN)) {
+        header_version_ = V2;
+      } else if (memcmp(buf_, PROXY_PROTO_V1_SIGNATURE, PROXY_PROTO_V1_SIGNATURE_LEN)) {
+        // It is not v2, and can't be v1, so no sense hanging around: it is invalid
+        throw EnvoyException("failed to read proxy protocol (exceed max v1 header len)");
       }
     }
 
-    // Read the data upto and including the line feed, if available, but not past it.
-    // This should never fail, as search_index_ - buf_off_ <= nread, so we're asking
-    // only for bytes we have already seen.
-    nread = recv(fd, buf_ + buf_off_, search_index_ - buf_off_, 0);
-    ASSERT(size_t(nread) == search_index_ - buf_off_);
+    if (header_version_ == V2) {
+      const int ver_cmd = buf_[PROXY_PROTO_V2_SIGNATURE_LEN];
+      if (((ver_cmd & 0xf0) >> 4) != PROXY_PROTO_V2_VERSION) {
+        throw EnvoyException("Unsupported V2 proxy protocol version");
+      }
+      if (buf_off_ < PROXY_PROTO_V2_HEADER_LEN) {
+        ssize_t lread;
+        ssize_t exp = PROXY_PROTO_V2_HEADER_LEN - buf_off_;
+        lread = os_syscalls.recv(fd, buf_ + buf_off_, exp, 0);
+        if (lread != exp) {
```

Comment:
When would this happen? Don't we check for enough data above? If you are concerned about syscall failure, is the "insufficient data" log message wrong? Same below?

Reply:
this is a race-condition check. above there was enough data, and, before we got to here, the socket was closed. e.g. RST, e.g. we ran out of kernel memory, e.g. some reached into /proc and closed it, etc.

in general, the only lread!=exp that is 'expected' is -1, socket closed, since the socket is blocking.

i suppose the message could say 'remote closed', so altered.